### ISSUE TITLE
Cabal Plugin is too strict for the newest cabal versions

### ISSUE THREAD

**Issue by fendor** [haskell/haskell-language-server#4401]
Generating a new package with cabal `3.12` via

```sh
$ mkdir test-hls-cabal
$ cd test-hls-cabal
$ cabal init --lib -n
```

Generates a `.cabal` file with `cabal-version: 3.12`, which `haskell-language-server-2.9.0.1` built with `ghc-9.6.5`, which is linked against `Cabal-3.10.3.0`. Consequentially, HLS highlights the whole `.cabal` file as incorrect, as it doesn't recognise `3.12` as a valid `cabal-version`.

![image](https://github.com/user-attachments/assets/1ac3087b-b220-496f-a721-413734d7db8c)

This is bad user-experience.

I am not sure whether we can fix that for good, as older HLS versions will never get linked against a newer `Cabal` version.
However, what we could do is to be more lenient when it comes to parsing `.cabal` files. 
We can scan the `.cabal` file for the line `cabal-version:` (`scanSpecVersion`) and if the version is not known to us, we parse the cabal file with the newest CabalSpecVersion available to the HLS binary.
The `cabal` features may still break if users actually need new cabal-the-spec-features, but for many future changes, we likely only display warnings and many other IDE features keep working.

The only real disadvantage is that we can't reliably report incorrect `cabal-version` numbers any more. Also, the proposed behaviour might be unexpected to users.

The last resort I can think of, is to simply not provide any IDE features if the cabal-version is seemingly too new for the Cabal version linked into HLS.

**Comment by fendor** [comment]
It seems to be more difficult than hoped to parse a `.cabal` file with a newer `cabal-version` in a lenient way. The `cabal-version` is parsed in the `FieldGrammar`, so we would have to basically copy-paste major parts of `Cabal-syntax` to not verify the provided `CabalSpecVersion` is not equal to the actual parsed ones.

Thus, the only way I can think of now is to basically overwrite the `cabal-version` field in-memory.

Ignoring `.cabal` files with a too new version seems to be the less hacky approach.

**Comment by michaelpj** [comment]
Yeah, I think just emitting a diagnostic saying that we don't know how to parse it is best? So pretty much the current behaviour, but with a nicer diagnostic that doesn't cover the whole file :sweat_smile:

**Comment by fridewald** [comment]
I'll give it a try

**Comment by fridewald** [comment]
There was an issue with handling errors in the cabal file if the error is in the first line. Cabal normally uses 1-based line numbering, but for the first line it can also return the line 0. Created a PR to safely fix this.

**Comment by fendor** [comment]
Fixed by #4425
